toolchain/parse/typed_nodes_test.cpp

```diff
@@ -0,0 +1,133 @@
+
+#include "toolchain/parse/typed_nodes.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <forward_list>
+
+#include "toolchain/diagnostics/mocks.h"
+#include "toolchain/lex/lex.h"
+#include "toolchain/lex/tokenized_buffer.h"
+#include "toolchain/parse/extract_file.h"
+#include "toolchain/parse/tree.h"
+
+namespace Carbon::Parse {
+namespace {
+
+// Check that each node kind defines a Kind member using the correct
+// NodeKind enumerator.
+#define CARBON_PARSE_NODE_KIND(Name) \
+  static_assert(Name::Kind == NodeKind::Name, #Name);
+#include "toolchain/parse/node_kind.def"
+
+class TypedNodeTest : public ::testing::Test {
+ protected:
+  auto GetSourceBuffer(llvm::StringRef t) -> SourceBuffer& {
+    CARBON_CHECK(fs_.addFile("test.carbon", /*ModificationTime=*/0,
+                             llvm::MemoryBuffer::getMemBuffer(t)));
+    source_storage_.push_front(std::move(
+        *SourceBuffer::CreateFromFile(fs_, "test.carbon", consumer_)));
+    return source_storage_.front();
+  }
+
+  auto GetTokenizedBuffer(llvm::StringRef t) -> Lex::TokenizedBuffer& {
+    token_storage_.push_front(
+        Lex::Lex(value_stores_, GetSourceBuffer(t), consumer_));
+    return token_storage_.front();
+  }
+
+  auto GetTree(llvm::StringRef t) -> Tree& {
+    tree_storage_.push_front(Tree::Parse(GetTokenizedBuffer(t), consumer_,
+                                         /*vlog_stream=*/nullptr));
+    return tree_storage_.front();
+  }
+
+  SharedValueStores value_stores_;
+  llvm::vfs::InMemoryFileSystem fs_;
+  std::forward_list<SourceBuffer> source_storage_;
+  std::forward_list<Lex::TokenizedBuffer> token_storage_;
+  std::forward_list<Tree> tree_storage_;
+  DiagnosticConsumer& consumer_ = ConsoleDiagnosticConsumer();
+};
+
+TEST_F(TypedNodeTest, Empty) {
+  auto* tree = &GetTree("");
+  auto file = File::Make(tree);
+
+  EXPECT_TRUE(tree->IsValid<FileStart>(file.start));
+  EXPECT_TRUE(tree->ExtractAs<FileStart>(file.start).has_value());
+  EXPECT_TRUE(tree->Extract(file.start).has_value());
+
+  EXPECT_TRUE(tree->IsValid<FileEnd>(file.end));
+  EXPECT_TRUE(tree->ExtractAs<FileEnd>(file.end).has_value());
+  EXPECT_TRUE(tree->Extract(file.end).has_value());
+
+  EXPECT_FALSE(tree->IsValid<FileEnd>(file.start));
+  EXPECT_FALSE(tree->ExtractAs<FileEnd>(file.start).has_value());
+}
+
+TEST_F(TypedNodeTest, Function) {
+  auto* tree = &GetTree(R"carbon(
+    fn F() {}
+    virtual fn G() -> i32;
+  )carbon");
+  auto file = File::Make(tree);
+
+  ASSERT_EQ(file.decls.size(), 2);
+
+  auto f_fn = tree->ExtractAs<FunctionDefinition>(file.decls[0]);
+  ASSERT_TRUE(f_fn.has_value());
+  auto f_sig = tree->Extract(f_fn->signature);
+  ASSERT_TRUE(f_sig.has_value());
+  EXPECT_FALSE(f_sig->return_type.has_value());
+  EXPECT_TRUE(f_sig->modifiers.empty());
+
+  auto g_fn = tree->ExtractAs<FunctionDecl>(file.decls[1]);
+  ASSERT_TRUE(g_fn.has_value());
+  EXPECT_TRUE(g_fn->return_type.has_value());
+  EXPECT_FALSE(g_fn->modifiers.empty());
```

Comment:
It'd be nice to have an example with multiple modifiers so we can check that we extract them in the right order (reversed from sibling order).